File: Reports/SummaryFlowReport.vb
'*********************************************************************************************
' STSimStockFlow: A SyncroSim Module for the ST-Sim Stocks and Flows Add-In.
'
' Copyright © 2007-2017 Apex Resource Management Solution Ltd. (ApexRMS). All rights reserved.
'
'*********************************************************************************************

Imports SyncroSim.Core
Imports SyncroSim.Core.Forms
Imports System.Globalization
Imports System.Reflection

<ObfuscationAttribute(Exclude:=True, ApplyToMembers:=False)>
Class SummaryFlowReport
    Inherits ExportTransformer

    Protected Overrides Sub Export(location As String, exportType As ExportType)
        Me.InternalExport(location, exportType, True)
    End Sub

    Friend Sub InternalExport(location As String, exportType As ExportType, ByVal showMessage As Boolean)

        Dim columns As ExportColumnCollection = Me.CreateColumnCollection()

        If (exportType = ExportType.ExcelFile) Then
            Me.ExcelExport(location, columns, Me.CreateReportQuery(False), "Flows")
        Else

            columns.Remove("ScenarioName")
            Me.CSVExport(location, columns, Me.CreateReportQuery(True))

            If (showMessage) Then
                FormsUtilities.InformationMessageBox("Data saved to '{0}'.", location)
            End If

        End If

    End Sub

    Private Function CreateColumnCollection() As ExportColumnCollection

        Dim c As New ExportColumnCollection()

        Dim dsterm As DataSheet = Me.Project.GetDataSheet(DATASHEET_TERMINOLOGY_NAME)
        Dim dstermSTSim As DataSheet = Me.Project.GetDataSheet("STSim_Terminology")
        Dim FlowUnits As String = GetTerminology(dsterm, STOCK_UNITS_COLUMN_NAME)
        Dim TimestepLabel As String = GetTimestepUnits(Me.Project)
        Dim PrimaryStratumLabel As String = Nothing
        Dim SecondaryStratumLabel As String = Nothing

        GetStratumLabelStrings(dstermSTSim, PrimaryStratumLabel, SecondaryStratumLabel)
        Dim TotalValue As String = String.Format(CultureInfo.InvariantCulture, "Total Value ({0})", FlowUnits)

        c.Add(New ExportColumn("ScenarioID", "Scenario ID"))
        c.Add(New ExportColumn("ScenarioName", "Scenario"))
        c.Add(New ExportColumn("Iteration", "Iteration"))
        c.Add(New ExportColumn("Timestep", TimestepLabel))
        c.Add(New ExportColumn("FromStratum", "From " & PrimaryStratumLabel))
        c.Add(New ExportColumn("FromSecondaryStratum", "From " & SecondaryStratumLabel))
        c.Add(New ExportColumn("FromStateClass", "From State Class"))
        c.Add(New ExportColumn("FromStock", "From Stock"))
        c.Add(New ExportColumn("TransitionType", "TransitionType"))
        c.Add(New ExportColumn("ToStratum", "To " & PrimaryStratumLabel))
        c.Add(New ExportColumn("ToStateClass", "To State Class"))
        c.Add(New ExportColumn("ToStock", "To Stock"))
        c.Add(New ExportColumn("FlowType", "Flow Type"))
        c.Add(New ExportColumn("Amount", TotalValue))

        c("Amount").DecimalPlaces = 2
        c("Amount").Alignment = Core.ColumnAlignment.Right

        Return c

    End Function

    Private Function CreateReportQuery(ByVal isCSV As Boolean) As String

        Dim ScenFilter As String = Me.CreateActiveResultScenarioFilter()

        If (isCSV) Then

            Return String.Format(CultureInfo.InvariantCulture,
                "SELECT " &
                "SF_OutputFlow.ScenarioID, " &
                "SF_OutputFlow.Iteration,  " &
                "SF_OutputFlow.Timestep,  " &
                "ST1.Name AS FromStratum, " &
                "SS1.Name AS FromSecondaryStratum, " &
                "SC1.Name AS FromStateClass, " &
                "STK1.Name AS FromStock, " &
                "STSim_TransitionType.Name AS TransitionType, " &
                "ST2.Name AS ToStratum, " &
                "SC2.Name AS ToStateClass, " &
                "STK2.Name AS ToStock, " &
                "SF_FlowType.Name as FlowType, " &
                "SF_OutputFlow.Amount " &
                "FROM SF_OutputFlow " &
                "INNER JOIN STSim_Stratum AS ST1 ON ST1.StratumID = SF_OutputFlow.FromStratumID " &
                "INNER JOIN STSim_Stratum AS ST2 ON ST2.StratumID = SF_OutputFlow.ToStratumID " &
                "LEFT JOIN STSim_SecondaryStratum AS SS1 ON SS1.SecondaryStratumID = SF_OutputFlow.FromSecondaryStratumID " &
                "INNER JOIN STSim_StateClass AS SC1 ON SC1.StateClassID = SF_OutputFlow.FromStateClassID " &
                "INNER JOIN STSim_StateClass AS SC2 ON SC2.StateClassID = SF_OutputFlow.ToStateClassID " &
                "INNER JOIN SF_StockType AS STK1 ON STK1.StockTypeID = SF_OutputFlow.FromStockTypeID " &
                "INNER JOIN SF_StockType AS STK2 ON STK2.StockTypeID = SF_OutputFlow.ToStockTypeID " &
                "INNER JOIN SF_FlowType ON SF_FlowType.FlowTypeID = SF_OutputFlow.FlowTypeID " &
                "LEFT JOIN STSim_TransitionType ON STSim_TransitionType.TransitionTypeID = SF_OutputFlow.TransitionTypeID " &
                "WHERE SF_OutputFlow.ScenarioID IN ({0})  " &
                "ORDER BY " &
                "SF_OutputFlow.ScenarioID, " &
                "SF_OutputFlow.Iteration, " &
                "SF_OutputFlow.Timestep, " &
                "ST1.Name, " &
                "SS1.Name, " &
                "SC1.Name, " &
                "STK1.Name, " &
                "STSim_TransitionType.Name, " &
                "ST2.Name, " &
                "SC2.Name, " &
                "STK2.Name, " &
                "SF_FlowType.Name",
                ScenFilter)

        Else

            Return String.Format(CultureInfo.InvariantCulture,
                "SELECT " &
                "SF_OutputFlow.ScenarioID, " &
                "SSim_Scenario.Name AS ScenarioName, " &
                "SF_OutputFlow.Iteration,  " &
                "SF_OutputFlow.Timestep,  " &
                "ST1.Name AS FromStratum, " &
                "SS1.Name AS FromSecondaryStratum, " &
                "SC1.Name AS FromStateClass, " &
                "STK1.Name AS FromStock, " &
                "STSim_TransitionType.Name AS TransitionType, " &
                "ST2.Name AS ToStratum, " &
                "SC2.Name AS ToStateClass, " &
                "STK2.Name AS ToStock, " &
                "SF_FlowType.Name as FlowType, " &
                "SF_OutputFlow.Amount " &
                "FROM SF_OutputFlow " &
                "INNER JOIN SSim_Scenario ON SSim_Scenario.ScenarioID = SF_OutputFlow.ScenarioID " &
                "INNER JOIN STSim_Stratum AS ST1 ON ST1.StratumID = SF_OutputFlow.FromStratumID " &
                "INNER JOIN STSim_Stratum AS ST2 ON ST2.StratumID = SF_OutputFlow.ToStratumID " &
                "LEFT JOIN STSim_SecondaryStratum AS SS1 ON SS1.SecondaryStratumID = SF_OutputFlow.FromSecondaryStratumID " &
                "INNER JOIN STSim_StateClass AS SC1 ON SC1.StateClassID = SF_OutputFlow.FromStateClassID " &
                "INNER JOIN STSim_StateClass AS SC2 ON SC2.StateClassID = SF_OutputFlow.ToStateClassID " &
                "INNER JOIN SF_StockType AS STK1 ON STK1.StockTypeID = SF_OutputFlow.FromStockTypeID " &
                "INNER JOIN SF_StockType AS STK2 ON STK2.StockTypeID = SF_OutputFlow.ToStockTypeID " &
                "INNER JOIN SF_FlowType ON SF_FlowType.FlowTypeID = SF_OutputFlow.FlowTypeID " &
                "LEFT JOIN STSim_TransitionType ON STSim_TransitionType.TransitionTypeID = SF_OutputFlow.TransitionTypeID " &
                "WHERE SF_OutputFlow.ScenarioID IN ({0})  " &
                "ORDER BY " &
                "SF_OutputFlow.ScenarioID, " &
                "SSim_Scenario.Name, " &
                "SF_OutputFlow.Iteration, " &
                "SF_OutputFlow.Timestep, " &
                "ST1.Name, " &
                "SS1.Name, " &
                "SC1.Name, " &
                "STK1.Name, " &
                "STSim_TransitionType.Name, " &
                "ST2.Name, " &
                "SC2.Name, " &
                "STK2.Name, " &
                "SF_FlowType.Name",
                ScenFilter)

        End If

    End Function

End Class

File: src/Reports/SummaryFlowReport.cs
// stsim-stockflow: SyncroSim Add-On Package (to stsim) for integrating stocks and flows into state-and-transition simulation models in ST-Sim.
// Copyright © 2007-2024 Apex Resource Management Solutions Ltd. (ApexRMS). All rights reserved.

using System.Reflection;
using System.Globalization;
using SyncroSim.Core;
using SyncroSim.Core.Forms;

namespace SyncroSim.STSimStockFlow
{
	[ObfuscationAttribute(Exclude=true, ApplyToMembers=false)]
	internal class SummaryFlowReport : ExportTransformer
	{
		protected override void Export(string location, ExportType exportType)
		{
			this.InternalExport(location, exportType, true);
		}

		internal void InternalExport(string location, ExportType exportType, bool showMessage)
		{
			ExportColumnCollection columns = this.CreateColumnCollection();

			if (exportType == ExportType.ExcelFile)
			{
				this.ExcelExport(location, columns, this.CreateReportQuery(false), "Flows");
			}
			else
			{
				columns.Remove("ScenarioName");
                this.CSVExport(location, columns, this.CreateReportQuery(true));

				if (showMessage)
				{
					FormsUtilities.InformationMessageBox("Data saved to '{0}'.", location);
				}
			}
		}

		private ExportColumnCollection CreateColumnCollection()
		{
			ExportColumnCollection c = new ExportColumnCollection();
			DataSheet dsterm = this.Project.GetDataSheet(Constants.DATASHEET_TERMINOLOGY_NAME);
			DataSheet dstermSTSim = this.Project.GetDataSheet(Constants.DATASHEET_STSIM_TERMINOLOGY);
			string FlowUnits = TerminologyUtilities.GetTerminology(dsterm, Constants.STOCK_UNITS_COLUMN_NAME);
			string TimestepLabel = TerminologyUtilities.GetTimestepUnits(this.Project);
			string PrimaryStratumLabel = null;
			string SecondaryStratumLabel = null;
			string TertiaryStratumLabel = null;

			TerminologyUtilities.GetStratumLabelStrings(dstermSTSim, ref PrimaryStratumLabel, ref SecondaryStratumLabel, ref TertiaryStratumLabel);
			string TotalValue = string.Format(CultureInfo.InvariantCulture, "Total Value ({0})", FlowUnits);

			c.Add(new ExportColumn("ScenarioID", "Scenario ID"));
			c.Add(new ExportColumn("ScenarioName", "Scenario"));
			c.Add(new ExportColumn("Iteration", "Iteration"));
			c.Add(new ExportColumn("Timestep", TimestepLabel));
			c.Add(new ExportColumn("FromStratum", "From " + PrimaryStratumLabel));
			c.Add(new ExportColumn("FromSecondaryStratum", "From " + SecondaryStratumLabel));
			c.Add(new ExportColumn("FromTertiaryStratum", "From " + TertiaryStratumLabel));
			c.Add(new ExportColumn("FromStateClass", "From State Class"));
			c.Add(new ExportColumn("FromStock", "From Stock"));
			c.Add(new ExportColumn("TransitionType", "TransitionType"));
			c.Add(new ExportColumn("ToStratum", "To " + PrimaryStratumLabel));
			c.Add(new ExportColumn("ToStateClass", "To State Class"));
			c.Add(new ExportColumn("ToStock", "To Stock"));
			c.Add(new ExportColumn("FlowGroup", "Flow Group"));
			c.Add(new ExportColumn("EndStratum", "End " + PrimaryStratumLabel));
			c.Add(new ExportColumn("EndSecondaryStratum", "End " + SecondaryStratumLabel));
			c.Add(new ExportColumn("EndTertiaryStratum", "End " + TertiaryStratumLabel));
			c.Add(new ExportColumn("EndStateClass", "End State Class"));
			c.Add(new ExportColumn("EndMinAge", "End Min Age"));

			c.Add(new ExportColumn("Amount", TotalValue));

			c["Amount"].DecimalPlaces = 2;
			c["Amount"].Alignment = ColumnAlignment.Right;

			return c;
		}

		private string CreateReportQuery(bool isCSV)
		{
			string ScenFilter = this.CreateActiveResultScenarioFilter();

            string Query =
                "SELECT " +
                "stsimsf_OutputFlow.ScenarioID, ";

            if (!isCSV)
            {
                Query += "core_Scenario.Name AS ScenarioName, ";
            }

            Query += string.Format(CultureInfo.InvariantCulture,
                "stsimsf_OutputFlow.Iteration,  " +
                "stsimsf_OutputFlow.Timestep,  " +
                "ST1.Name AS FromStratum, " +
                "SS1.Name AS FromSecondaryStratum, " +
                "TS1.Name AS FromTertiaryStratum, " +
                "SC1.Name AS FromStateClass, " +
                "STK1.Name AS FromStock, " +
                "stsim_TransitionType.Name AS TransitionType, " +
                "ST2.Name AS ToStratum, " +
                "SC2.Name AS ToStateClass, " +
                "STK2.Name AS ToStock, " +
                "stsimsf_FlowGroup.Name as FlowGroup, " +
                "ST3.Name AS EndStratum, " +
                "SS2.Name AS EndSecondaryStratum, " +
                "TS2.Name AS EndTertiaryStratum, " +
                "SC3.Name AS EndStateClass, " +
                "stsimsf_OutputFlow.EndMinAge, " +
                "stsimsf_OutputFlow.Amount " +
                "FROM stsimsf_OutputFlow " +
                "INNER JOIN core_Scenario ON core_Scenario.ScenarioID = stsimsf_OutputFlow.ScenarioID " +
                "INNER JOIN stsim_Stratum AS ST1 ON ST1.StratumID = stsimsf_OutputFlow.FromStratumID " +
                "INNER JOIN stsim_Stratum AS ST2 ON ST2.StratumID = stsimsf_OutputFlow.ToStratumID " +
                "LEFT JOIN stsim_Stratum AS ST3 ON ST3.StratumID = stsimsf_OutputFlow.EndStratumID " +
                "LEFT JOIN stsim_SecondaryStratum AS SS1 ON SS1.SecondaryStratumID = stsimsf_OutputFlow.FromSecondaryStratumID " +
                "LEFT JOIN stsim_SecondaryStratum AS SS2 ON SS2.SecondaryStratumID = stsimsf_OutputFlow.EndSecondaryStratumID " +
                "LEFT JOIN stsim_TertiaryStratum AS TS1 ON TS1.TertiaryStratumID = stsimsf_OutputFlow.FromTertiaryStratumID " +
                "LEFT JOIN stsim_TertiaryStratum AS TS2 ON TS2.TertiaryStratumID = stsimsf_OutputFlow.EndTertiaryStratumID " +
                "INNER JOIN stsim_StateClass AS SC1 ON SC1.StateClassID = stsimsf_OutputFlow.FromStateClassID " +
                "INNER JOIN stsim_StateClass AS SC2 ON SC2.StateClassID = stsimsf_OutputFlow.ToStateClassID " +
                "LEFT JOIN stsim_StateClass AS SC3 ON SC3.StateClassID = stsimsf_OutputFlow.EndStateClassID " +
                "INNER JOIN stsimsf_StockType AS STK1 ON STK1.StockTypeID = stsimsf_OutputFlow.FromStockTypeID " +
                "INNER JOIN stsimsf_StockType AS STK2 ON STK2.StockTypeID = stsimsf_OutputFlow.ToStockTypeID " +
                "INNER JOIN stsimsf_FlowGroup ON stsimsf_FlowGroup.FlowGroupID = stsimsf_OutputFlow.FlowGroupID " +
                "LEFT JOIN stsim_TransitionType ON stsim_TransitionType.TransitionTypeID = stsimsf_OutputFlow.TransitionTypeID " +
                "WHERE stsimsf_OutputFlow.ScenarioID IN ({0})  " +
                "ORDER BY " +
                "stsimsf_OutputFlow.ScenarioID, ", ScenFilter);

            if (!isCSV)
            {
                Query += "core_Scenario.Name, ";
            }
             
            Query +=       
                "stsimsf_OutputFlow.Iteration, " + 
                "stsimsf_OutputFlow.Timestep, " + 
                "ST1.Name, " + 
                "SS1.Name, " + 
                "TS1.Name, " + 
                "SC1.Name, " + 
                "STK1.Name, " + 
                "stsim_TransitionType.Name, " + 
                "ST2.Name, " + 
                "SC2.Name, " +
                "STK2.Name, " + 
                "stsimsf_FlowGroup.Name, " + 
                "ST3.Name, " +
                "SS2.Name, " +
                "TS2.Name, " +
                "SC3.Name, " +
                "stsimsf_OutputFlow.EndMinAge";

            return Query;
		}
	}
}
